src/CanvasRenderingContext2d.cc

```diff
@@ -310,17 +371,35 @@ Context2d::setFillRule(v8::Local<v8::Value> value) {
 
 void
 Context2d::fill(bool preserve) {
+  cairo_pattern_t *new_pattern;
   if (state->fillPattern) {
-    cairo_set_source(_context, state->fillPattern);
+    if (state->globalAlpha < 1) {
+      new_pattern = create_transparent_pattern(state->fillPattern, state->globalAlpha);
+      cairo_set_source(_context, new_pattern);
+      if (new_pattern != state->fillPattern) {
+        cairo_pattern_destroy(new_pattern);
+      }
```

Comment:
Looks great! Last comment... do I have this right: `new_pattern == state->fillPattern` only if `create_transparent_pattern` fails? If so, then I'd consider changing this (and the equivalent other blocks of code) to:

```cpp
new_pattern = create_transparent_pattern(state->fillPattern, state->globalAlpha);
if (new_pattern == state->fillPattern) {
  // failed to allocate; Nan::ThrowError has already been called, so return from this fn.
  return;
}
cairo_set_source(_context, new_pattern);
cairo_pattern_destroy(new_pattern);
```

I think that clarifies the flow a bit, and explicitly stops execution if we've reached an error state.

Reply:
and what about gradient? gradient has a similiar case, a slower chance to fail since i have no idea how you can create gradients type that are not linear or radial, but if this happen,still we are in a sort of error state, or at least rendering something different from what is expected.
Do we want to return also in that case?

Reply:
I think you can do the same thing...

```cpp
new_pattern = create_transparent_gradient(state->fillGradient, state->globalAlpha);
if (new_pattern == state->fillGradient) return;
cairo_pattern_set_filter(new_pattern, state->patternQuality);
cairo_set_source(_context, new_pattern);
```

Also, since you're now bailing out in case of error, you might as well return `NULL` from `create_transparent_gradient`/`create_transparent_pattern` in case of error, and then just check if `new_pattern == NULL`.

Reply:
but can i return null from a function that expect a pointer to cairo_pattern_t?

Reply:
Yep!